src/main/java/server/http/RequestHandler.java

```diff
@@ -37,35 +38,77 @@ void handle() {
         });
     }
 
-    private void handleInternal() throws IOException, MalformedRequestException {
+    private void handleInternal() throws IOException {
         try (OutputStream output = socket.getOutputStream();
              PrintWriter printOutput = new PrintWriter(output);
              BufferedReader input = new BufferedReader(new InputStreamReader(socket.getInputStream()))) {
 
-            final RequestLine requestLine = parseRequestLine(input.readLine());
-
             final ResponseWriter responseWriter = new ResponseWriter(output, printOutput);
 
-            if (!isMethodAllowed(requestLine)) {
-                LOGGER.info("Sending method not allowed response...");
-
-                responseWriter.writeMethodNotAllowed();
-            } else if (!resourceExists(requestLine)) {
-                LOGGER.info("Sending not found response...");
-
-                responseWriter.writeNotFound();
-            } else {
-                LOGGER.info("Resource found!");
+            respond(input.readLine(), responseWriter);
 
-                responseWriter.writeFile(resolvedPath(requestLine));
-            }
         } finally {
             socket.close();
 
             LOGGER.info("Socket closed, transmission over.");
         }
     }
 
+    private void respond(final String line, final ResponseWriter responseWriter){
+        try {
+            final RequestLine requestLine = parseRequestLine(line);
+            ProcessChain.of(requestLine, responseWriter)
+                    .process(this::checkMethodAllowed)
+                    .process(this::checkResourceExists)
+                    .process(this::checkReadAllowed)
+                    .process(this::writeOk);
+        } catch (MalformedRequestException e) {
+            responseWriter.writeBadRequest();
+            LOGGER.log(Level.SEVERE, e.getMessage(), e);
+        } catch (Exception e) {
+            responseWriter.writeInternalServerError();
+            LOGGER.log(Level.SEVERE, e.getMessage(), e);
+        }
+    }
+
+    private boolean writeOk(final RequestLine requestLine, final ResponseWriter responseWriter) {
+        LOGGER.info("Resource found!");
+        try {
+            responseWriter.writeFile(resolvedPath(requestLine));
+        } catch (IOException e) {
+            throw new RuntimeException(e.getMessage(), e);
```

Comment:
Please avoid throwing a naked `RuntimeException`. Introducing some kind of wrapper exception would be great.